Merge Triton ScaledMM as epilogue to MM template

PaulZhang12 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #150415
* __->__ #150045

Previously, scaled_mm's (FP8 matmul) Triton lowering for inductor was in a separate template. This PR consolidates that lowering into the mm template, with an added epilogue to deal with multiplying the scales. This paves the way for future scaled variants of BMM, Grouped GEMM in inductor.

Currently, there is still a separate template for TMA+persistent version of scaled_mm. The current mm lowering has a separate template for TMA + Persistent version. Will hopefully consolidate the extra scaled_mm TMA+persistent template when the consolidation for the mm template is done.
TODO: Consolidate TMA+Persistent logic into 1 template and remove separate scaled_mm TMA template


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

PaulZhang12 committed:
Consolidate mm_scaled into mm template

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Consolidate mm_scaled into mm template"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Consolidate mm_scaled into mm template"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Consolidate mm_scaled into mm template"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

drisspg commented on torch/_inductor/ir.py:
```diff
@@ -3950,14 +3951,17 @@ def freeze_layout_with_exact_strides(  # type: ignore[no-untyped-def]
     def is_zero_elements(self):  # type: ignore[no-untyped-def]
         return V.graph.sizevars.is_expr_static_and_true(sympy.Eq(self.get_numel(), 0))
 
+    def is_scalar(self):  # type: ignore[no-untyped-def]
```
Is this needed for per tensor scaling?

PaulZhang12 replied:
When per tensor scaling is a scalar, size is [] as is strides, which doesn't match the 2D shape of the resulting matmul.

I guess we can also pad the strides/size of the scalar input to be [1, 1]? Not sure what is the preferred option here

eellison replied:
Can you expand the scalar instead ? if we're going to make a change to the loading code, let's do that as a separate pr (but i dont think we need to).

drisspg reviewed (approved):
I think looks good, can you run the inducto scaled_mm tests locally and paste the output here for provenance since we dont have h100 testing in CI

PaulZhang12 commented:
@drisspg sounds good! Is that just `python test/inductor/test_fp8.py -- TestFP8Lowering  2>&1 | tee ~/test.log`?

eellison commented on torch/_inductor/kernel/mm.py:
```diff
@@ -124,7 +131,11 @@
         {{load_input("B", "b", ("idx_m", "idx_n"), mask=None if EVEN_K else "b_mask", indent_width=8)}}
-        acc += tl.dot(a, b, allow_tf32=ALLOW_TF32)
+
+        if USE_FAST_ACCUM:
+            acc = tl.dot(a, b, acc, allow_tf32=ALLOW_TF32, out_dtype=ACC_TYPE)
+        else:
+            acc += tl.dot(a, b, allow_tf32=ALLOW_TF32, out_dtype=ACC_TYPE)
```
nit: could we do 
```
{% if USE_FAST_ACCUM %}
...
```

this makes the generated output code easier to read

eellison commented on torch/_inductor/kernel/mm.py:
```diff
@@ -741,6 +963,116 @@ def tuned_sparse_semi_structured_mm(
+    
+    scale_a, scale_b = realize_inputs(scale_a, scale_b)
+
+    # input_nodes: tuple[Any, ...]
+    # workaround for Inductor not supporting optional tensor input arguments
+    # if bias is None:
```
intended ?

eellison commented on torch/_inductor/kernel/mm.py:
```diff
@@ -741,6 +963,116 @@ def tuned_sparse_semi_structured_mm(
+        else:
+            triton_input_nodes = input_nodes
+
+        if use_triton_tma_template(mat_a, mat_b) and bias is None:
+            for config in scaled_persistent_mm_configs(m, n, k):
+                kwargs = scaled_mm_options(
+                    config, m, n, k, layout, scale_a, scale_b, use_fast_accum, device_tma=True
+                )
+                scaled_mm_device_tma_template.maybe_append_choice(
+                    choices,
+                    input_nodes=triton_input_nodes,
+                    layout=layout,
+                    workspace_arg=get_tma_workspace_arg(
+                        num_tma_descriptors=2,
+                        device=mat_a.get_device(),
+                    ),
+                    **kwargs,
+                )
+        else:
+            for config in scaled_mm_configs(m, n, k):
+                if k == 16 and config.kwargs["BLOCK_M"] >= 64:
```
I dont think these were intended to be mutually exclusive. the tma is intended to be autotune'd with the other configs no ?

drisspg replied:
So I intially added these as mutually exclusive but yeah we likely want to merge before flipping the switch

eellison commented on torch/_inductor/utils.py:
```diff
@@ -1376,6 +1376,10 @@ def _is_tma_compatible(x: IRNode) -> bool:
         if transposed:
             inner_dim = layout.size[0]
+        
+        if dtype == torch.float8_e4m3fn and inner_dim < 32:
```
inner_dim can be sympy expr so you should use statically_known_lt

eellison commented on torch/_inductor/kernel/mm.py:
```diff
@@ -741,6 +963,116 @@ def tuned_sparse_semi_structured_mm(
+        else:
+            triton_input_nodes = input_nodes
+
+        if use_triton_tma_template(mat_a, mat_b) and bias is None:
```
why are we skipping on bias here ?

PaulZhang12 replied:
I believe it's because the original scaled_mm code required a separate template for bias and tma, and there wasn't a template combining bias + tma? I think this might be safe to take out, thoughts @drisspg

drisspg replied:
Yup, agree with Paul here

eellison commented on torch/_inductor/kernel/mm.py:
```diff
@@ -360,6 +561,27 @@ def bias_addmm(inp, mat1, mat2, *, out=None, alpha=1, beta=1):
 
+def check_supported_striding(mat_a, mat_b) -> None:
+    def is_row_major(stride) -> bool:
+        return stride[1] == 1
```
would also expect these to use the sympy related functions

PaulZhang12 replied:
Sure, can I ask why this is exactly?

eellison replied:
We do a bit more in our attempts to prove equality, using information from shape env: https://github.com/pytorch/pytorch/blob/01cb3519b3eb10fb007dd64241c679663eb7dfe6/torch/_inductor/sizevars.py#L164-L169

It probably less here in this case but is more in keeping with codebase..

drisspg commented:
@PaulZhang12 Yup thats the one

eellison commented on torch/_inductor/ir.py:
```diff
@@ -3549,6 +3549,7 @@ def indexer(index):  # type: ignore[no-untyped-def]
             for idx, stride, sz in zip(index, self.stride, self.size):
                 if sz != 1:
                     result = result + idx * stride
+
```
nit: formatting changes

eellison reviewed (commented):
looking good, a few comments

PaulZhang12 committed:
Update on "Consolidate mm_scaled into mm template"


Previously, scaled_mm's (FP8 matmul) Triton lowering for inductor was in a separate template. This PR consolidates that lowering into the mm template, with an added epilogue to deal with multiplying the scales. This paves the way for future scaled variants of BMM, Grouped GEMM in inductor.

Currently, there is still a separate template for TMA+persistent version of scaled_mm. The current mm lowering has a separate template for TMA + Persistent version. Will hopefully consolidate the extra scaled_mm TMA+persistent template when the consolidation for the mm template is done.
TODO: Consolidate TMA+Persistent logic into 1 template and remove separate scaled_mm TMA template


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Consolidate mm_scaled into mm template"


Previously, scaled_mm's (FP8 matmul) Triton lowering for inductor was in a separate template. This PR consolidates that lowering into the mm template, with an added epilogue to deal with multiplying the scales. This paves the way for future scaled variants of BMM, Grouped GEMM in inductor.

Currently, there is still a separate template for TMA+persistent version of scaled_mm. The current mm lowering has a separate template for TMA + Persistent version. Will hopefully consolidate the extra scaled_mm TMA+persistent template when the consolidation for the mm template is done.
TODO: Consolidate TMA+Persistent logic into 1 template and remove separate scaled_mm TMA template


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

PaulZhang12 commented:
@drisspg https://pastebin.com/BgYsCXHE is the result of running `python test/inductor/test_fp8.py -- TestFP8Lowering` on h100 machine. There are some minor precision issues with FAST_ACCUM=True, but I believe trunk also has these issues

PaulZhang12 committed:
Update on "Consolidate mm_scaled into mm template"


Previously, scaled_mm's (FP8 matmul) Triton lowering for inductor was in a separate template. This PR consolidates that lowering into the mm template, with an added epilogue to deal with multiplying the scales. This paves the way for future scaled variants of BMM, Grouped GEMM in inductor.

Currently, there is still a separate template for TMA+persistent version of scaled_mm. The current mm lowering has a separate template for TMA + Persistent version. Will hopefully consolidate the extra scaled_mm TMA+persistent template when the consolidation for the mm template is done.
TODO: Consolidate TMA+Persistent logic into 1 template and remove separate scaled_mm TMA template


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Consolidate mm_scaled into mm template"


Previously, scaled_mm's (FP8 matmul) Triton lowering for inductor was in a separate template. This PR consolidates that lowering into the mm template, with an added epilogue to deal with multiplying the scales. This paves the way for future scaled variants of BMM, Grouped GEMM in inductor.

Currently, there is still a separate template for TMA+persistent version of scaled_mm. The current mm lowering has a separate template for TMA + Persistent version. Will hopefully consolidate the extra scaled_mm TMA+persistent template when the consolidation for the mm template is done.
TODO: Consolidate TMA+Persistent logic into 1 template and remove separate scaled_mm TMA template


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

PaulZhang12 commented:
@pytorchbot merge

PaulZhang12 committed:
Update on "Consolidate mm_scaled into mm template"


Previously, scaled_mm's (FP8 matmul) Triton lowering for inductor was in a separate template. This PR consolidates that lowering into the mm template, with an added epilogue to deal with multiplying the scales. This paves the way for future scaled variants of BMM, Grouped GEMM in inductor.

Currently, there is still a separate template for TMA+persistent version of scaled_mm. The current mm lowering has a separate template for TMA + Persistent version. Will hopefully consolidate the extra scaled_mm TMA+persistent template when the consolidation for the mm template is done.
TODO: Consolidate TMA+Persistent logic into 1 template and remove separate scaled_mm TMA template


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

PaulZhang12 commented:
@pytorchbot merge

PaulZhang12 commented:
@pytorchmergebot merge -f "Unrelated build failure"

PaulZhang12 commented:
@pytorchbot revert -m "Need to add PR 150415 fixes for internal merge" -c ghfirst

PaulZhang12 commented:
Reopened in https://github.com/pytorch/pytorch/pull/150441, ghstack creates multiple PRs
